mlir/lib/Dialect/Rock/Transforms/Regularize.cpp

```diff
@@ -147,187 +146,452 @@ struct RegularizeGenericRewritePattern
+  auto maybeBufferWriters = bufferDeps.getWriters(allocOp);
+  auto maybeBufferReaders = bufferDeps.getReaders(allocOp);
+  if (!maybeBufferWriters)
+    return allocOp.emitOpError(
+        "couldn't find an operation that writes this buffer");
+  if (!maybeBufferReaders)
+    return allocOp.emitError(
+        "couldn't find an operation that reads this buffer");
+  SmallVector<OpOperand *> bufferWriters = std::move(*maybeBufferWriters);
+  SmallVector<OpOperand *> bufferReaders = std::move(*maybeBufferReaders);
+  if (bufferWriters.size() != 1) {
+    allocOp->getParentOp()->dump();
+    return allocOp.emitError(
+               "expected one operation to write this buffer, but there are ")
+           << bufferWriters.size();
+  }
+
+  OpOperand *writer = bufferWriters[0];
+  bool isInputFusion = state.inputFusionWrites.contains(writer);
+  if (isInputFusion)
+    return pushTransformsToInputFusionReaders(allocOp, bufferDeps, writer,
+                                              state, rw);
+  if (!state.outputFusionWrites.contains(writer)) {
+    return allocOp.emitOpError("couldn't find the writer of a buffer in the "
+                               "set of analyzed fusable writes. See ")
+           << bufferWriters.front()->get();
+  }
+
+  OpOperand *transformingReader = nullptr;
+  for (OpOperand *reader : bufferReaders) {
+    Value readArg = reader->get();
+    if (!readArg.getDefiningOp<rock::TransformOp>()) {
+      continue;
     }
-    return lres;
+    if (transformingReader && transformingReader != reader)
+      return isolateMultipleTransformingReaders(
+          allocOp, writer->getOwner(), bufferDeps, bufferReaders, state, rw);
+    transformingReader = reader;
   }
-};
+  // It's the simple case, do nothing
+  if (!transformingReader)
+    return success();
+
+  if (transformingReader && bufferReaders.size() > 1) {
+    LLVM_DEBUG(llvm::dbgs()
+               << "Adding copies to partially intercept a transform stack\n");
+    return isolateMultipleTransformingReaders(
+        allocOp, writer->getOwner(), bufferDeps, bufferReaders, state, rw);
+  }
+
+  LLVM_DEBUG(llvm::dbgs() << "Processing output fusion read via "
+                          << transformingReader->get() << " (argument "
+                          << transformingReader->getOperandNumber() << " of "
+                          << *transformingReader->getOwner() << ") via "
+                          << allocOp << "\n");
+
+  return regularizeTransformStack(allocOp, transformingReader, bufferDeps,
+                                  state, rw);
+}
+
+LogicalResult findFusionRoots(func::FuncOp kernel,
+                              const BufferDependencyAnalysis &bufferDeps,
+                              TransformPushState &state) {
+  bool foundFusionRoot = false;
+  kernel.walk([&](Operation *op) {
+    llvm::TypeSwitch<Operation *>(op)
+        .Case(
+            [&](memref::AllocOp allocOp) { state.worklist.push_back(allocOp); })
```

Comment:
This will capture all the `memref::AllocOp` s in the function. yes ? 
If so why do I see worklist.push_back of allocs in fusionwrites collection functions ?

Reply:
The extra push_back()s are guards against bugs - if you've already corrected the issue, you should be able to re-run the analysis again and have it be a noop. I remember leaving comments about that